### PR TITLE
dbt 1.3.0 upgrade

### PR DESCRIPTION
**Pull request by mdesmet**
## Checklist

- [x] I have run this code in development and it appears to resolve the stated issue
- [x] This PR includes tests, or tests are not required/relevant for this PR
- [x] `README.md` updated and added information about my change
- [x] I have run `changie new` to [create a changelog entry](https://github.com/starburstdata/dbt-trino/blob/master/CONTRIBUTING.md#Adding-CHANGELOG-Entry)

### CODE REVIEW

File: dbt/include/trino/macros/materializations/incremental.sql [hunk 1]
```diff
@@ -1,35 +1,104 @@
...
+       in order to guarantee consistent inputs to both statements.
+
+       If we are running a single statement (MERGE or INSERT alone),
+       we can save the model query definition as a view instead,
+       for faster overall incremental processing.
+  #} */
+
+  {% if language == 'sql' and (strategy in ('default', 'append', 'merge') or (unique_key is none)) %}
+    {{ return('view') }}
+  {% else %}  {#--  play it safe -- #}
```
Comment (hovaesco) [inline dbt/include/trino/macros/materializations/incremental.sql:16]: remove the comment

File: dbt/include/trino/macros/materializations/incremental.sql [hunk 2]
```diff
@@ -1,35 +1,104 @@
...
-  {% elif strategy == 'delete+insert' %}
-    {% do return(get_delete_insert_merge_sql(target_relation, tmp_relation, unique_key, dest_columns)) %}
-  {% elif strategy == 'merge' %}
-    {% do return(get_merge_sql(target_relation, tmp_relation, unique_key, dest_columns, predicates=none)) %}
+{% materialization incremental, adapter='trino', supported_languages=['sql'] -%}
+
+  {#-- Set vars --#}
+  {%- set full_refresh_mode = (should_full_refresh()) -%}
+  {%- set language = model['language'] -%}
+  {#-- TODO: figure out why we need to add table explicitly, otherwise persist_docs doesn't know the type --#}
```
Comment (hovaesco) [inline dbt/include/trino/macros/materializations/incremental.sql:26]: remove the comment

File: dbt/include/trino/macros/materializations/incremental.sql [hunk 3]
```diff
@@ -61,43 +130,43 @@
...
         {% endif %}
 
         {{ sql_header if sql_header is not none }}
 
-        merge into {{ target }} as DBT_INTERNAL_DEST {{ "\n" }}using {{ source }} as DBT_INTERNAL_SOURCE on
-        {{ predicates | join(' and
-        ') }}
-
-        {%- if unique_key %}
+        merge into {{ target }} as DBT_INTERNAL_DEST
```
Comment (hovaesco) [inline dbt/include/trino/macros/materializations/incremental.sql:160]: It may break the SQL formatting, please validate

File: dbt/include/trino/macros/materializations/incremental.sql [hunk 4]
```diff
@@ -61,43 +130,43 @@
...
+            using {{ source }} as DBT_INTERNAL_SOURCE
+            on {{ predicates | join(' and ') }}
 
+        {% if unique_key %}
         when matched then update set
             {% for column_name in update_columns -%}
                 {{ column_name }} = DBT_INTERNAL_SOURCE.{{ column_name }}
-                {%- if not loop.last %},
-            {% endif %}
+                {%- if not loop.last %}, {%- endif %}
```
Comment (hovaesco) [inline dbt/include/trino/macros/materializations/incremental.sql:168]: The same with formatting

File: dbt/include/trino/macros/utils/array_construct.sql [hunk 5]
```diff
@@ -0,0 +1,7 @@
+{% macro trino__array_construct(inputs, data_type) -%}
+    {%- if not inputs -%}
+    NULL
+    {%- else -%}
+    ARRAY[ {{ inputs|join(' , ') }} ]
```
Comment (hovaesco) [inline dbt/include/trino/macros/utils/array_construct.sql:5]: `ARRAY` lowercase to keep consistency

**Review by hovaesco — COMMENTED** [review]
Looks good, please squash dbt 1.3.0 upgrade related commits together.

File: dbt/include/trino/macros/materializations/incremental.sql (hunk 1 above)
Comment (mdesmet) [reply dbt/include/trino/macros/materializations/incremental.sql:16]: decided to keep it the same as in dbt-core to limit differences and make further upgrades as smooth as possible.

File: dbt/include/trino/macros/materializations/incremental.sql (hunk 2 above)
Comment (mdesmet) [reply dbt/include/trino/macros/materializations/incremental.sql]: Removed comment

File: dbt/include/trino/macros/materializations/incremental.sql (hunk 3 above)
Comment (mdesmet) [reply dbt/include/trino/macros/materializations/incremental.sql:159]: Validated, it is the same as core dbt.

File: dbt/include/trino/macros/materializations/incremental.sql (hunk 4 above)
Comment (mdesmet) [reply dbt/include/trino/macros/materializations/incremental.sql:167]: Validated, it is the same as core dbt.

File: dbt/include/trino/macros/utils/array_construct.sql (hunk 5 above)
Comment (mdesmet) [reply dbt/include/trino/macros/utils/array_construct.sql]: converted to lowercase

**Review by hovaesco — APPROVED** [review]

File: README.md [hunk 6]
```diff
@@ -227,6 +227,8 @@ select * from {{ ref('events') }}
 
 Use the `+on_schema_change` property to define how dbt-trino should handle column changes. See [dbt docs](https://docs.getdbt.com/docs/building-a-dbt-project/building-models/configuring-incremental-models#what-if-the-columns-of-my-incremental-model-change).
 
+Set the `+views_enabled` to `false` if your connector doesn't support views.
```
Comment (findinpath) [inline README.md:230]: Please add more details about where the setting needs to be made.

File: dbt/adapters/trino/connections.py [hunk 7]
```diff
@@ -460,11 +459,17 @@ def add_query(self, sql, auto_begin=True, bindings=None, abridge_sql_log=False):
             )
 
         if cursor is None:
-            raise dbt.exceptions.RuntimeException(
+            conn = self.get_thread_connection()
```
Comment (findinpath) [inline dbt/adapters/trino/connections.py:462]: Is the exception handling related to dbt 1.3 upgrade?
If not, let's do it separately to leave more time to discuss it in detail.

File: dbt/include/trino/macros/materializations/incremental.sql [hunk 8]
```diff
@@ -1,35 +1,104 @@
...
-    {% do exceptions.raise_compiler_error(invalid_strategy_msg) %}
+{% macro get_incremental_tmp_relation_type(strategy, unique_key, language) %}
+
+  /* {#
+       If we are running multiple statements (DELETE + INSERT),
+       we must first save the model query results as a temporary table
+       in order to guarantee consistent inputs to both statements.
+
+       If we are running a single statement (MERGE or INSERT alone),
+       we can save the model query definition as a view instead,
```
Comment (findinpath) [inline dbt/include/trino/macros/materializations/incremental.sql:9]: as a view instead (if views are available for the Trino connector corresponding to the target model)

File: dbt/include/trino/macros/materializations/incremental.sql [hunk 9]
```diff
@@ -61,43 +130,43 @@
     )
 {%- endmacro %}
 
-{% macro trino__get_merge_sql(target, source, unique_key, dest_columns, predicates=none) -%}
+{% macro trino__get_merge_sql(target, source, unique_key, dest_columns, predicates) -%}
```
Comment (findinpath) [inline dbt/include/trino/macros/materializations/incremental.sql:133]: do we have any differences here compared to the default dbt implementation?

i remember we had found a bug in a previous Trino implementation which caused to rewrite the macro in Trino

cc @hovaesco

File: tests/functional/adapter/utils/test_utils.py [hunk 10]
```diff
@@ -33,6 +37,22 @@ class TestAnyValue(BaseAnyValue):
     pass
 
 
+# Not working because of https://github.com/trinodb/trino/issues/13
+@pytest.mark.skip(reason="No way to concat an array with null or empty array")
```
Comment (findinpath) [inline tests/functional/adapter/utils/test_utils.py:41]: The test is failing for the specific use case 

```
select 2 as id, {{ array_append(array_construct([]), 4) }} as array_col
```


Let's rather override the test case logic and omit this specific use case.

File: dbt/adapters/trino/connections.py [hunk 11]
```diff
@@ -419,8 +419,8 @@ def open(cls, connection):
             max_attempts=credentials.retries,
             isolation_level=IsolationLevel.AUTOCOMMIT,
             source="dbt-trino",
+            verify=credentials.cert,
```
Comment (findinpath) [inline dbt/adapters/trino/connections.py:421]: we can do this in a separate PR

File: .changes/unreleased/Breaking Changes-20221014-185350.yaml [hunk 12]
```diff
@@ -0,0 +1,9 @@
+kind: Breaking Changes
+body: By default incremental materializations will use views if all incremental operations
```
Comment (findinpath) [inline .changes/unreleased/Breaking Changes-20221014-185350.yaml:2]: Isn't this related to [Upgrade dbt to 1.3](https://github.com/starburstdata/dbt-trino/pull/161/commits/ba64e8f8dd87ca4fd485539858c4df38568d3760) ?

File: dbt/include/trino/macros/materializations/incremental.sql (hunk 9 above)
Comment (mdesmet) [reply dbt/include/trino/macros/materializations/incremental.sql:133]: We had two issues:

* We would not use false as a merge match condition
* Ambiguous columns in merge match condition

File: dbt/adapters/trino/connections.py (hunk 7 above)
Comment (mdesmet) [reply dbt/adapters/trino/connections.py:462]: This is something we already discussed before, @hovaesco  had this issue and it seems resolved exactly like this in the dbt's snowflake adapter. There are probably some conditions within dbt that may result in having no thread connection. It is unclear how to reproduce it but probably something really happening in the field.

File: .changes/unreleased/Breaking Changes-20221014-185350.yaml (hunk 12 above)
Comment (mdesmet) [reply .changes/unreleased/Breaking Changes-20221014-185350.yaml:2]: I will move it to this one and also squash the incremental refactor into this commit.

**Review by findinpath — APPROVED** [review]
